[VIDEO-3095] Facilitate removal of setup-gstreamer by splitting gstreamer action into pieces

btgoodwin opened the pull request:
VIDEO-3095:

The goal of the ticket and its subtasks is to make it more straight-forward to directly use our cerbero build action(s) with other projects vs. how we were using `setup-gstreamer` to download and install MSIs.

1. `setup-gstreamer`: Configures variables, identifies version numbers, etc. and looks for each installer on S3.  If installers are found, the build is skipped and the installers are downloaded and ran.  Otherwise, the build is run and the locally-held installers are installed.
2. `build-gstreamer`: Formerly build-gstreamer-for-windows is now stripped to the studs of bootstrapping the environment, running the build, and uploading to s3 (optionally).
3. `install-gstreamer`: Installs the MSIs from either the local '-path' or downloads and then installs from the '-url' variables.

External projects should point to `setup-gstreamer`.

Testing:

1. [Job where not-found-build-install worked](https://github.com/blinemedical/cerbero/actions/runs/13929835453/attempts/1)
2. [Job run where find-download-install worked](https://github.com/blinemedical/cerbero/actions/runs/13929835453/job/38988886526)
3. [Job where force-build-install worked](https://github.com/blinemedical/cerbero/actions/runs/13930896358/job/38989469923)

> Note to future self (and others): in order to "test" 1->2, go to the successful (1) and re-run all jobs.  This will ensure that the commit sha used for the upload in (1) and the sha picked up by (2) are identical because (1) was initiated using a merge request SHA, not the branch head.  If you go to Actions and manually run the action, that's basically (3) because it's going to grab the branch head.

btgoodwin committed:
action: split gstreamer action into pieces

1. setup-gstreamer:
   Configures variables, identifies version numbers, etc. and looks for
   each installer on S3.  If installers are found, the build is skipped
   and the installers are downloaded and ran.  Otherwise, the build is
   run and the locally-held installers are installed.
2. build-gstreamer:
   Formerly build-gstreamer-for-windows is now stripped to to the
   studs of bootstrapping the environment, running the build, and
   uploading to s3 (optionally).
3. install-gstreamer:
   Installs the MSIs from either the local '-path' or downloads
   and then installs from the '-url' variables.

btgoodwin commented:
I'm not sure what is the better play here:

1. Merge the concepts from cerbero-action to here.
2. Merge the concepts from here to cerbero-action.

(1) Is probably the least amount of work.
(2) Helps us have a place to direct users of setup-gstreamer assuming we also validate the Debian Linux side of the build, have it produce and upload tarballs.

jawilson commented on .github/actions/install-gstreamer/action.yml:
```diff
@@ -0,0 +1,101 @@
+      shell: msys2 -eo pipefail {0}
+      run: |
+        echo "Error: Cannot specify both development-installer-url and development-installer-path"
+        exit 2
```
We should also error if neither is specified, or if the file specified by `inputs.runtime-installer-path` doesn't exist.

btgoodwin replied:
Fixed in db2f1f37.

jawilson commented on .github/actions/setup-gstreamer/action.yml:
```diff
@@ -0,0 +1,177 @@
+          bucket=$(echo ${download_path} | cut -d/ -f3)
+          prefix=$(echo ${download_path} | cut -d/ -f4-)
+
+          if [ "$runtime_msi" = "null" ]; then
+            echo "Checking for runtime installer under s3://${bucket}/${prefix}/"
+            runtime_msi=$(aws s3api list-objects-v2 --bucket "${bucket}" --prefix "${prefix}/" --query 'Contents[].Key' | jq -r 'if type!="array" then [] else . end | sort_by(length) | map(select(endswith("'"${RUNTIME_MSI_FILENAME}"'"))) | .[0]')
+            if [ "$runtime_msi" != "null" ]; then
+              echo "runtime-url=$(aws s3 presign s3://${bucket}/${runtime_msi})" | tee -a $GITHUB_OUTPUT
+            fi
+          fi
+
+          if [ "$dev_msi" = "null" ]; then
+            echo "Checking for development installer under s3://${bucket}/${prefix}/"
+            dev_msi=$(aws s3api list-objects-v2 --bucket "${bucket}" --prefix "${prefix}/" --query 'Contents[].Key' | jq -r 'if type!="array" then [] else . end | sort_by(length) | map(select(endswith("'"${DEV_MSI_FILENAME}"'"))) | .[0]')
```
`RUNTIME_MSI_FILENAME` and `DEV_MSI_FILENAME` are both undefined here, so I think it's just finding the first build to end with an empty string, which would be all of them. Each time you push to your branch a full build should be done because it's a whole new [`cerbero_short_sha`](https://github.com/blinemedical/cerbero/blob/18ec28f35700510ce693d36dc13e261f75758fc1/.github/actions/setup-gstreamer/action.yml#L101). I think you just need to pass them in like the old action did:
```yaml
      env:
        RUNTIME_MSI_FILENAME: ${{ steps.version-info.outputs.runtime-msi-filename }}
        DEV_MSI_FILENAME: ${{ steps.version-info.outputs.dev-msi-filename }}
```

btgoodwin replied:
Fixed in 8e562417.

jawilson commented on .github/workflows/windows.yml:
```diff
@@ -81,6 +81,6 @@ jobs:
 
       - name: Annotate workflow run with GStreamer version
-        if: steps.build-gstreamer.outputs.gstreamer-version
+        if: steps.setupo-gstreamer.outputs.gstreamer-version
```
Typo, which definitely made me chuckle a bit
```suggestion
        if: steps.setup-gstreamer.outputs.gstreamer-version
```

btgoodwin replied:
I'm amazed nothing caught that as an error/warning given the variety of other things I've read.

Fixed in 636a6dea.

jawilson commented on .github/actions/setup-gstreamer/action.yml:
```diff
@@ -0,0 +1,177 @@
+        dev-msi-filename: ${{ steps.version-info.outputs.dev-msi-filename }}
+        force: ${{ inputs.force }}
+        cleanup: ${{ inputs.cleanup }}
+        no-cache: ${{ inputs.no-cache }}
```
The necessary args to upload a build to S3 (`s3-download-paths` and `s3-upload-path`) aren't specified so the build is never uploaded.
That said, it feels to me like uploading builds to S3 should be moved into this action since it's also the one attempting to download them. That would also make the `build-gstreamer` action _only_ build GStreamer. Additionally, because we're now directly using the installers created by `build-gstreamer` there's no need to generate any pre-signed URLs after uploading the installers.

jawilson commented on .github/actions/build-gstreamer/action.yml:
Comments in this file from lines that haven't been changed (and thus GH won't let me comment on):

- The following inputs aren't used in this action:
  - `cerbero-repo`
  - `cerbero-ref`
  - `s3-download-paths`
  - `force`
  - `inputs.cleanup` - technically used in `outputs.cerbero-path` but that's deceptive
- And if you move uploading to S3 to the `setup-gstreamer` action these will also be unused (you can also get rid of the `*-installer-url` outputs):
  - `s3-upload-path`
  - `runtime-msi-filename`
  - `dev-msi-filename`
- `gstreamer-version` should be inferred within this action with `./cerbero-uninstalled packageinfo gstreamer-1.0 | awk '/Version:/ {print $2}'` even if it duplicates from `setup-gstreamer`

jawilson commented on .github/actions/setup-gstreamer/action.yml:
```diff
@@ -0,0 +1,177 @@
+        development-installer-path: ${{ steps.build.outputs.devel-installer-path }}
+
+    - name: Cleanup Cerbero
+      shell: msys2 -eo pipefail {0}
```
Should be gated by the `cleanup` input:
```suggestion
    - name: Cleanup Cerbero
      if: ${{ always() && inputs.cleanup == 'true' && steps.get-cerbero.outcome != 'skipped' }}
      shell: msys2 -eo pipefail {0}
```

jawilson reviewed (changes requested):
I do prefer option (2). I don't think we have to validate the Debian side right away, but worth throwing a ticket in. I'm not sure the S3 stuff will be very widely used, but it's very much optional (just don't specify any `s3-*` inputs).

btgoodwin committed:
action: pr commento abouto setupo typ (oh)

btgoodwin committed:
action: quit on error if neither path, url are set; validate path

Complete the XOR by also throwing an error if neither path or url
is set for each dev and runtime input.  Also if the path is set,
verify the file actually exists (or error out).

btgoodwin committed:
action: restore [DEV|RUNTIME]_MSI_FILENAME vars.

These were accidentally deleted and are necessary for correctly
searching S3 for release installers.

btgoodwin committed:
action: 'build' only builds; 'setup' performs upload if build ran

PR suggestion to make 'build' only perform building, no uploading,
and for the setup action to perform the upload, if the build is
a success, because the setup action also performs the download
if the installer already exists.

btgoodwin committed:
action: removing/reworking unused variables

btgoodwin committed:
action: post-action 'cleanup' tied to vars, always()

btgoodwin commented:
Sorry for the couple force commits here.  Some of these builds are actually going through which is helping to identify things that ...sure would have been nice to have been caught by a linter, like step names changing and the references not being updated.

btgoodwin commented:
Aaaand once more to rebase against 1.24-lldc since it received changes this morning.

btgoodwin committed:
action: fix path selection going into install-gstreamer call

btgoodwin commented:
Putting this back to draft.  Current issues:

1. It supposedly runs the upload step but doesn't upload anything.
2. The upload filenames don't contain the hash, which I thought previously they did for the sake of making some sense out of which-build-is-which.
3. If I can get the upload to work, I can finally verify the "don't build just download" logic paths.

btgoodwin committed:
action: fix upload path to include commit sha

btgoodwin committed:
action: adding --no-progress to aws s3 cp commands

Getting rid of superfluous log spam.

btgoodwin commented:
This _seems ready now_ @jawilson.

jawilson commented on .github/actions/build-gstreamer/action.yml:
```diff
@@ -326,33 +202,6 @@ runs:
         find ~+ -maxdepth 1 -name 'gstreamer-1.0-msvc*.msi' -exec sh -c 'echo runtime-msi=$(cygpath -aw "{}") >> $GITHUB_OUTPUT' \;
         find ~+ -maxdepth 1 -name 'gstreamer-1.0-devel*.msi' -exec sh -c 'echo dev-msi=$(cygpath -aw "{}") >> $GITHUB_OUTPUT' \;
 
```
We should annotate the workflow here:

```suggestion

    - name: Annotate workflow run with GStreamer version
      run: |
        echo "::notice title=Built GStreamer version ${{ steps.cerbero-config.outputs.version }}::Created installer with Cerbero short sha ${{ steps.cerbero-config.outputs.cerbero-sha }}"

```

jawilson commented on .github/actions/install-gstreamer/action.yml:
```diff
@@ -0,0 +1,127 @@
+      run: |
+        $msi_path = "$(( $pwd ))\gstreamer-runtime.msi"
+        Invoke-WebRequest -Uri "${{ inputs.runtime-installer-url }}" -OutFile "$msi_path"
+        echo "msi-path=$msi_path" | Out-File -FilePath $Env:GITHUB_OUTPUT -Append -Encoding utf8
```
Wouldn't hurt to annotate here:


```suggestion
        echo "msi-path=$msi_path" | Out-File -FilePath $Env:GITHUB_OUTPUT -Append -Encoding utf8
        echo "::notice title=GStreamer runtime installer downloaded::Downloaded GStreamer runtime installer from ${{ inputs.runtime-installer-url }}"
```

jawilson commented on .github/actions/install-gstreamer/action.yml:
```diff
@@ -0,0 +1,127 @@
+      run: |
+        $msi_path = "$(( $pwd ))\gstreamer-development.msi"
+        Invoke-WebRequest -Uri "${{ inputs.development-installer-url }}" -OutFile "$msi_path"
+        echo "msi-path=$msi_path" | Out-File -FilePath $Env:GITHUB_OUTPUT -Append -Encoding utf8
```
Same as above
```suggestion
        echo "msi-path=$msi_path" | Out-File -FilePath $Env:GITHUB_OUTPUT -Append -Encoding utf8
        echo "::notice title=GStreamer development installer downloaded::Downloaded GStreamer development installer from ${{ inputs.development-installer-url }}"  
```

jawilson commented on .github/actions/build-gstreamer/action.yml:
```diff
@@ -37,150 +30,32 @@ inputs:
-    value: ${{ inputs.cleanup != 'true' && 'cerbero' }}
+    required: true
+    default: ''
+outputs:
```
Outputting the GStreamer version here is useful, as an input, it isn't used at all.

```suggestion
outputs:
  gstreamer-version:
    description: 'GStreamer version built'
    value: ${{ steps.cerbero-config.outputs.version }}
```

jawilson commented on .github/workflows/windows.yml:
```diff
@@ -81,6 +81,6 @@ jobs:
-        if: steps.build-gstreamer.outputs.gstreamer-version
+        if: steps.setup-gstreamer.outputs.gstreamer-version
         run: |
           echo "::notice title=Built GStreamer version ${{ steps.build-gstreamer.outputs.gstreamer-version }}::Created installer with Cerbero short sha ${{ steps.build-gstreamer.outputs.cerbero-short-sha }}"
```
This is no longer needed with my other annotation recommendations.
```suggestion
```

jawilson reviewed (changes requested):
Ran and re-ran and behavior was as expected! Just a few comments to address and this will be good to go.

btgoodwin committed:
pr comment: move gstreamer-version to output

btgoodwin committed:
pr: update annotations

btgoodwin committed:
pr: fixing typo in previous commit

btgoodwin committed:
pr: fixing pr suggestion

Was missing shell

jawilson commented on .github/actions/setup-gstreamer/action.yml:
```diff
@@ -0,0 +1,209 @@
+            runtime_msi=$(aws s3api list-objects-v2 --bucket "${bucket}" --prefix "${prefix}/" --query 'Contents[].Key' | jq -r 'if type!="array" then [] else . end | sort_by(length) | map(select(endswith("'"${RUNTIME_MSI_FILENAME}"'"))) | .[0]')
+            if [ "$runtime_msi" != "null" ]; then
+              aws s3 cp --no-progress "s3://${bucket}/${runtime_msi}" "cerbero/$(basename ${runtime_msi})"
+              echo "runtime-path=cerbero/$(basename ${runtime_msi})" | tee -a $GITHUB_OUTPUT
```
```suggestion
              s3_uri="s3://${bucket}/${runtime_msi}"
              aws s3 cp --no-progress "${s3_uri}" "cerbero/$(basename ${runtime_msi})"
              echo "runtime-path=cerbero/$(basename ${runtime_msi})" | tee -a $GITHUB_OUTPUT
              echo "::notice title=GStreamer Runtime Installer Downloaded From S3::Location: ${s3_uri}"
```

jawilson commented on .github/actions/setup-gstreamer/action.yml:
```diff
@@ -0,0 +1,209 @@
+            echo "Checking for development installer under s3://${bucket}/${prefix}/"
+            dev_msi=$(aws s3api list-objects-v2 --bucket "${bucket}" --prefix "${prefix}/" --query 'Contents[].Key' | jq -r 'if type!="array" then [] else . end | sort_by(length) | map(select(endswith("'"${DEV_MSI_FILENAME}"'"))) | .[0]')
+            if [ "$dev_msi" != "null" ]; then
+              aws s3 cp --no-progress "s3://${bucket}/${dev_msi}" "cerbero/$(basename ${dev_msi})"
+              echo "dev-path=cerbero/$(basename ${dev_msi})" | tee -a $GITHUB_OUTPUT
```
```suggestion
              s3_uri="s3://${bucket}/${dev_msi}"
              aws s3 cp --no-progress "${s3_uri}" "cerbero/$(basename ${dev_msi})"
              echo "dev-path=cerbero/$(basename ${dev_msi})" | tee -a $GITHUB_OUTPUT
              echo "::notice title=GStreamer Development Installer Downloaded From S3::Location: ${s3_uri}"
```

jawilson reviewed (approved):
I missed one last set of annotations and then we should be good!

jawilson reviewed (changes requested):
Bah, VS Code PR interface messed with me again, meant to request changes on that last one.

btgoodwin committed:
pr suggestions

Co-authored-by: Jeff Wilson <jeffery.wilson@laerdal.com>

jawilson reviewed (approved):
Excellent work!
